test/freenet/support/FieldsDurationTest.java

```diff
@@ -0,0 +1,36 @@
+package freenet.support;
+
+import freenet.config.Dimension;
+import junit.framework.TestCase;
+
+/**
+ * Tests parsing of duration value.
+ */
+public class FieldsDurationTest extends TestCase {
+
+    /**
+     * Duration input with and without various d|h|min|s.
+     */
+    private static final String[] durations = { "2d", "3h", "20m", "56s", "1h30m" };
```

Comment:
You can use a parametrized test to make this show all failing tests, and to make it trivial to add more tests. https://github.com/junit-team/junit4/wiki/Parameterized-tests

Reply:
Please do not write parameterized tests.

- Parameterized tests can’t tell you any more than that two values didn’t match but they offer no context. Tests are supposed to be documentation, by means of the name of the test method and the used variable names.
- It’s easy to mix up values in long lists and so a broken test or broken code can stay undetected for longer. That’s easily avoidable by creating a method for each value pair.
- They encourage people to add more tests that add no value but increase the number of tests.

Reply:
For a parser like this, I disagree.

If you use the parametrized tests runner, then you always have pairs of {input, output}, ideally each on its own line, and you can annotate the variables to give good output. For a test with lots of edge-case-values like a parser, that’s a really efficient way to test, and clear to read.

For a parser, parametrized tests are documentation by example. Due to the conciseness of one input-output pair per line, it’s easier to spot mistakes, because the values are not buried in several lines of repetitive ceremony.

And having the test-values in one block makes it easier to spot unnecessary duplication of tested logic, because you can grasp all the test values with one glance.

To show that by way of example:

```java
    @Parameters(name = "{index}: duration string: {0}, millis: {1}")
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][] {
                 // single units
                 { "2d", 2 * 24 * 3600 * 1000},
                 { "3h", 3 * 3600 * 1000},
                 { "20m", 20 * 60 * 1000},
                 { "56s", 56 * 1000},
                 // mixed units
                 { "1h30m", 5400 * 1000}
           });
    }
```